plotly_resampler/aggregation/aggregation_interface.py

```diff
@@ -4,169 +4,260 @@
+            there are gaps.
+        idxs: np.ndarray
+            The index array. This is relevant aggregators that perform data point
+            selection (e.g., max, min, etc.) - this array will be expanded with the
+            same indices where there are gaps.
+
+        Returns
+        -------
+        Tuple[np.ndarray, np.ndarray]
+            The expanded y_agg array and the expanded idxs array respectively.
+
+        """
+        gap_mask = AbstractAggregator._get_gap_mask(x_agg)
+        if gap_mask is None:
+            # no gaps are found, nothing to do
+            return y_agg, idxs
+
+        # An array filled with 1s and 2s, where 2 indicates a large gap mask
+        # (i.e., that index will be repeated twice)
+        repeats = np.ones(x_agg.shape, dtype="int") + gap_mask
+
+        # use the repeats to expand the idxs, and agg_y array
+        idx_exp_nan = np.repeat(idxs, repeats)
+        y_agg_exp_nan = np.repeat(y_agg, repeats)
+
+        # only float arrays can contain NaN values
+        if issubclass(y_agg_exp_nan.dtype.type, np.integer) or issubclass(
+            y_agg_exp_nan.dtype.type, np.bool_
+        ):
+            y_agg_exp_nan = y_agg_exp_nan.astype("float")
+
+        # Set the NaN values
+        # We add the gap index offset (via the np.arange) to the indices to account for
+        # the repeats (i.e., expanded y_agg array).
+        y_agg_exp_nan[np.where(gap_mask)[0] + np.arange(gap_mask.sum())] = None
 
-    def _supports_dtype(self, s: pd.Series):
+        return y_agg_exp_nan, idx_exp_nan
+
+    @staticmethod
+    def _supports_dtype(arr: np.ndarray, dtype_regex_list: Optional[List[str]] = None):
         # base case
-        if self.dtype_regex_list is None:
+        if dtype_regex_list is None:
             return
 
-        for dtype_regex_str in self.dtype_regex_list:
-            m = re.compile(dtype_regex_str).match(str(s.dtype))
+        for dtype_regex_str in dtype_regex_list:
+            m = re.compile(dtype_regex_str).match(str(arr.dtype))
             if m is not None:  # a match is found
                 return
         raise ValueError(
-            f"{s.dtype} doesn't match with any regex in {self.dtype_regex_list}"
+            f"{arr.dtype} doesn't match with any regex in {dtype_regex_list}"
         )
 
-    @staticmethod
-    def _calc_med_diff(s: pd.Series) -> Tuple[float, np.ndarray]:
-        # ----- divide and conquer heuristic to calculate the median diff ------
-        # remark: thanks to the prepend -> s_idx_diff.shape === len(s)
-        siv = s.index.values
-        s_idx_diff = np.diff(s.index.values, prepend=siv[0])
 
-        # To do so - use a quantile-based (median) approach where we reshape the data
-        # into `n_blocks` blocks and calculate the min
-        n_blcks = 128
-        if s.shape[0] > 5 * n_blcks:
-            blck_size = s_idx_diff.shape[0] // n_blcks
+class DataAggregator(AbstractAggregator, ABC):
```

Comment:
This inheritance is a bit of a red flag - I don't want to propose design changes, but using composition or just simple old normal functions might be cleaner (for the static parts in the parent class).

Reply:
Fully agree - we should favor composition. Will leave this currently as is. However, we should refactor this ASAP - gap handling should be decoupled from this & added via composition (as such, users can define their own gap handling methodology)

Reply:
Also, gap handling should be decoupled from the DataAggregator like classes & added via composition (as such, users can define their own gap handling methodology)

Reply:
Will create an issue for this!

Reply:
https://github.com/predict-idlab/plotly-resampler/issues/189

Feel free to propose design changes over there